src/compute/src/render/flat_map.rs

```diff
@@ -51,51 +59,54 @@ where
-                            let mut time = time.clone();
-                            *time.event_time() = event_time;
-                            Ok((row, time, diff))
-                        }
-                        Err((e, event_time, diff)) => {
-                            // Copy the whole time, and re-populate event time.
-                            let mut time = time.clone();
-                            *time.event_time() = event_time;
-                            Err((e, time, diff))
+                // Draw rows out of the table func evaluation.
+                // Consolidate them as we acculumate them, and when we are "full" apply
+                // the MFP and record and perhaps consolidate the results.
+                for (output_row, r) in output_rows {
+                    table_func_output.push((output_row, r));
+                    if table_func_output.len() == table_func_output.capacity() {
+                        // Consolidate the collection of things we will append to `input_row`.
+                        differential_dataflow::consolidation::consolidate(&mut table_func_output);
+                        if table_func_output.len() > table_func_output.capacity() / 2 {
+                            drain_through_mfp(
+                                &input_row,
+                                &mut time,
+                                &diff,
+                                &mut datums_mfp,
+                                &table_func_output,
+                                &mfp_plan,
+                                &until,
+                                &mut oks_output,
+                                &mut err_output,
+                            );
+
+                            // Double the capacity of the table func output buffer.
+                            table_func_output.clear();
+                            table_func_output.reserve(2 * table_func_output.capacity());
                         }
-                    })
-                    .collect::<Vec<_>>()
+                    }
+                }
+
+                // Drain any straggler extensions.
+                if !table_func_output.is_empty() {
+                    drain_through_mfp(
+                        &input_row,
+                        &mut time,
+                        &diff,
+                        &mut datums_mfp,
+                        &table_func_output,
+                        &mfp_plan,
+                        &until,
+                        &mut oks_output,
+                        &mut err_output,
+                    );
+                }
+                // About to be dropped, but here to remind us in case we ever re-use it.
+                table_func_output.clear();
+
+                let oks = oks_output.into_iter().map(Ok);
+                let err = err_output.into_iter().map(Err);
+                oks.chain(err).collect::<Vec<_>>()
```

Comment:
Follow-up: Rewrite the operator as a rc operator to immediately produce the outputs instead of muxing/demuxing.